File: 백준/Silver/10816. 숫자 카드 2/숫자 카드 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;


public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());
        int arr[] = new int[n];

        st = new StringTokenizer(br.readLine(), " ");
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(arr);

        int m = Integer.parseInt(br.readLine());
        st = new StringTokenizer(br.readLine(), " ");
        for (int i = 0; i < m; i++) {
            int key = Integer.parseInt(st.nextToken());
            sb.append(upperBinary(arr, key) - lowerBinary(arr, key)).append(" ");
        }
        System.out.println(sb);
    }

    static int upperBinary(int arr[], int key) {    //  초과
        int low = 0;
        int high = arr.length;

        while (low < high) {
            int mid = (low + high) / 2;

            if (key < arr[mid]) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }

    static int lowerBinary(int arr[], int key) {    //  이상
        int low = 0;
        int high = arr.length;

        while (low < high) {
            int mid = (low + high) / 2;

            if (key <= arr[mid]) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
}

File: 백준/Silver/10816. 숫자 카드 2/숫자 카드 2.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.HashMap
import java.util.StringTokenizer

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val n = br.readLine().toInt()
    val map = HashMap<Int, Int>()

    var st = StringTokenizer(br.readLine())
    repeat(n) {
        val input = st.nextToken().toInt()
        map[input] = map.getOrDefault(input, 0) + 1
    }

    val m = br.readLine().toInt()

    st = StringTokenizer(br.readLine())
    val sb = StringBuilder()
    repeat(m) {
        val num = st.nextToken().toInt()
        sb.append(map.getOrDefault(num, 0)).append(" ")
    }

    bw.write(sb.toString())
    bw.flush()
    bw.close()
}